### PR TITLE
Add ExternalDeploymentTriggerDagRunOperator

### PR DESCRIPTION
**Pull request by odaneau-astro**
Adds an ExternalDeploymentTriggerDagRunOperator to start new Dag runs on external Airflow deployments through the Airflow REST API.

### CODE REVIEW

**Comment by rajaths010494** [comment]
We need to defer and trigger the dag from Trigger not in execute method.
Adding a reference for the same sensor and how it is implemented https://github.com/astronomer/astronomer-providers/blob/main/astronomer/providers/core/sensors/external_task.py#L38-L62.

**Review by phanikumv — COMMENTED** [review]
@odaneau-astro please let us know if you need any help to move this forward

**Comment by odaneau-astro** [comment]
Thanks for the offer. Time was missing over the last few weeks. 
The code is mostly done, I'm adding tests and will be able to open the PR sometime this week.

**Comment by odaneau-astro** [comment]
I have considered using the Airflow API Python client at https://github.com/apache/airflow-client-python
Chose to build a custom Client instead since not all endpoints are currently supported.
Curious to have feedback on this.

File: astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from airflow.operators.python import PythonOperator
+from airflow.utils.log.secrets_masker import mask_secret
+from airflow.utils.timezone import datetime
+
+from astronomer.providers.core.operators.external_trigger_dagrun import (
+    ExternalDeploymentTriggerDagRunOperator,
+)
+
+DEPLOYMENT_CONN_ID = os.getenv("ASTRO_DEPLOYMENT_CONN_ID", "")
+ASTRO_AUTH_CONN_ID = os.getenv("ASTRO_AUTH_CONN_ID", "")
```
Comment (pankajkoti) [inline astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py:17]: If we have one, can we please add a link to the documentation in a comment here describing how to setup an Astro auth connection?

File: astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+
+with DAG(
+    dag_id="example_external_deployment_trigger_dag_operator",
+    start_date=datetime(2022, 1, 1),
+    schedule_interval=None,
+    catchup=False,
+    default_args=default_args,
+    tags=["example", "async", "http"],
```
Comment (pankajkoti) [inline astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py:59]: Do we need an `http` tag here?

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+        endpoint = DAGRUN_ENDPOINT.format(dag_id=dag_id, run_id=run_id)
+        hook = HttpHook(method="GET", http_conn_id=http_conn_id)
+
+        return hook.url_from_endpoint(endpoint=endpoint)
+
+
+class ExternalDeploymentTriggerDagRunOperator(BaseOperator):
+    """
+        External Deployment Trigger Dag Run Operator Make HTTP call to trigger a Dag Run and poll for the response state of externally
+        deployed DAG run to complete. The host should be external deployment url, header must contain access token
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:56]:
```suggestion
        External Deployment Trigger Dag Run Operator makes an HTTP call to trigger a run of an externally deployed Dag and poll for its status asynchronously. The host should be the external deployment URL and the header must contain a valid access token
```

Comment (pankajkoti) [reply astronomer/providers/core/operators/external_trigger_dagrun.py:56]: Might need to check the indentation afterwards if you consider the suggestion :)

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+            Dag run conf is immutable and will not be reset on rerun of an existing dag run.
+            When reset_dag_run=False and dag run exists, DagRunAlreadyExists will be raised.
+            When reset_dag_run=True and dag run exists, existing dag run will be cleared to rerun.
+        :param wait_for_completion: Whether wait for dag run completion. (default: False)
+        :param poke_interval: Poke interval to check dag run status when wait_for_completion=True.
+            (default: 60)
+        :param allowed_states: List of allowed states, default is ``['success']``.
+        :param failed_states: List of failed or dis-allowed states, default is ``None``.
+
+    Changed in version 2.1.3: 'queued' is added as a possible value.
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:87]: possible value for what (do we mean here)?

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+    template_fields: Sequence[str] = (
+        "headers",
+        "trigger_dag_id",
+        "trigger_run_id",
+        "execution_date",
+        "conf",
+    )
+    template_fields_renderers = {"headers": "json", "data": "py", "conf": "py"}
+    ui_color = "#7352ba"
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:98]: Will it be possible to add a comment here regarding what colour this hash code corresponds to :)?
Looks like `Blue Marguerite` https://colors.dopely.top/color-pedia/7352ba

Also curious to understand if there was some significance behind choosing this colour. Is it that we use it in some other places too?

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+            This is useful when backfill or rerun an existing dag run.
+            This only resets (not recreates) the dag run.
+            Dag run conf is immutable and will not be reset on rerun of an existing dag run.
+            When reset_dag_run=False and dag run exists, DagRunAlreadyExists will be raised.
+            When reset_dag_run=True and dag run exists, existing dag run will be cleared to rerun.
+        :param wait_for_completion: Whether wait for dag run completion. (default: False)
+        :param poke_interval: Poke interval to check dag run status when wait_for_completion=True.
+            (default: 60)
+        :param allowed_states: List of allowed states, default is ``['success']``.
+        :param failed_states: List of failed or dis-allowed states, default is ``None``.
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:85]: Isn't the default state `[State.FAILED]`?

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+            raise AirflowException("conf parameter should be JSON Serializable") from e
+
+    def execute(self, context: Context) -> None:
+        """
+        Trigger a DAG run on an external deployment.
+        :param context: Airflow context
+        """
+        execution_date = parse_execution_date(self.execution_date)
+
+        api = AirflowApiClient(
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:157]:
```suggestion
        api_client = AirflowApiClient(
```

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+                    dag_id=self.trigger_dag_id, execution_date=execution_date
+                ).run_id
+
+            api.clear_dag_run(dag_id=self.trigger_dag_id, run_id=self.trigger_run_id)
+            dag_run = api.get_dag_run(dag_id=self.trigger_dag_id, run_id=self.trigger_run_id)
+
+        if dag_run is None:
+            raise RuntimeError("The dag_run should be set here!")
+        logging.warning(type(dag_run))
+        logging.warning(inspect.getmembers(dag_run))
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:194]: Why are we setting the log level to `warning` here?

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+        )
+
+
+def parse_dag_run(json_dict: dict[str, Any]) -> DagRun:
+    """
+    Parse JSON dictionary to DagRun object.
+
+    :param json_dict: Dictionary containing DagRun properties to parse
+    """
+    if (data_interval_start := json_dict.get("data_interval_start")) and (
```
Comment (pankajkoti) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:262]: Sadly, the [setup.cfg](https://github.com/astronomer/astronomer-providers/blob/main/setup.cfg#L32) mentions `python>=3.7`. So the usage of walrus operators may not be supported on `3.7`

If we wish to continue using this, we might need to raise an exception while using the operator is Python version `<3.8` or refactor this bit here to also support `3.7 `

File: astronomer/providers/core/triggers/external_dagrun.py
Comment (pankajkoti) [inline astronomer/providers/core/triggers/external_dagrun.py:1]: This file name seems to be inconsistent with the operator file name. In general, we try to keep it consistent.

File: tests/core/operators/test_external_dagrun.py [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+#             conf=self.CONF,
+#             execution_date=datetime(2023, 1, 1),
+#             unpause_dag=True,
+#             reset_dag_run=True,
+#             wait_for_completion=True,
+#             headers=self.HEADERS,
+#             poke_interval=5,
+#         )
+#         with pytest.raises(ValueError):
+#             operator.execute_complete(context={}, event=mock_resp)
```
Comment (pankajkoti) [inline tests/core/operators/test_external_dagrun.py:409]: Do we not need this block? If not, can we remove the commented block?

**Review by pankajkoti — COMMENTED** [review]
Looks great efforts to me 👏🏽 . Have some comments inline. Also, did we get a chance to test the operator and trigger such an externally deployed DAG to validate the operator end to end?

**Comment by odaneau-astro** [comment]
I'm sorry I let this PR open for so long. Was working on PS engagements and forgot about it.
I'll address all the comments and fix the failed CI checks this week.

File: astronomer/providers/core/operators/external_trigger_dagrun.py (hunk 4 above)
Comment (odaneau-astro) [reply astronomer/providers/core/operators/external_trigger_dagrun.py:87]: Removed it. It was copy-pasted by mistake.

File: astronomer/providers/core/operators/external_trigger_dagrun.py (hunk 5 above)
Comment (odaneau-astro) [reply astronomer/providers/core/operators/external_trigger_dagrun.py:98]: removed it completely. no need to be fancy with custom UI colors, the default one works.

**Comment by odaneau-astro** [comment]
@phanikumv : Fixed all the errors and comments.
Somehow mypy still throws an error on  astronomer/providers/microsoft/azure/hooks/wasb.py:59.
It is not something I've changed with this PR. Let me know what you want to do with it.

File: astronomer/providers/core/operators/external_trigger_dagrun.py [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+    Operator link for ExternalDeploymentTriggerDagRunOperator.
+
+    It allows users to access DAG triggered by task using ExternalDeploymentTriggerDagRunOperator.
+    """
+
+    name = "Triggered DAG"
+
+    def get_link(self, operator: BaseOperator, *, ti_key: TaskInstanceKey) -> str:
+        """Provide link to browse externally triggered Dag run."""
+        # execution_date = XCom.get_value(ti_key=ti_key, key=XCOM_EXECUTION_DATE_ISO)
```
Comment (pankajastro) [inline astronomer/providers/core/operators/external_trigger_dagrun.py:40]: do we need this or we can remove?

**Comment by pankajastro** [comment]
Look like maybe some more tests require.

**Comment by phanikumv** [comment]
@odaneau-astro this PR is going on since quite some time. not sure whether we still need it?

**Comment by phanikumv** [comment]
Closing as there is no activity on this since 2 months

**Pull request closed without merging**